File: Data Structures & Algorithms/count-number-of-islands/submission-0.go
func dfs(i int, j int, vis [][]int, grid [][]byte, m int, n int) {
	if i >= m || i < 0 || j >= n || j < 0 || grid[i][j] == '0' || vis[i][j] == 1 {
		return
	}

	vis[i][j] = 1

	dfs(i+1, j, vis, grid, m, n)
	dfs(i-1, j, vis, grid, m, n)
	dfs(i, j+1, vis, grid, m, n)
	dfs(i, j-1, vis, grid, m, n)
}

func numIslands(grid [][]byte) int {
	m, n := len(grid), len(grid[0])

	vis := make([][]int, m)
	for i := 0; i < m; i++ {
		vis[i] = make([]int, n)
		for j := 0; j < n; j++ {
			vis[i][j] = -1
		}
	}

	ans := 0

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == '1' && vis[i][j] == -1 {
				dfs(i, j, vis, grid, m, n)
				ans++
			}
		}
	}

	return ans
}

File: Data Structures & Algorithms/max-area-of-island/submission-0.cpp
class Solution {
public:
    int dfs(int i, int j, vector<vector<int>>& grid, vector<vector<int>>& vis) {
        if(i < 0 || i >= grid.size() || j < 0 || j >= grid[0].size() || grid[i][j] == 0 || vis[i][j] == 1)
            return 0;
        vis[i][j] = 1;
        return 1 + dfs(i + 1, j, grid, vis) + 
               dfs(i - 1, j, grid, vis) + 
               dfs(i, j + 1, grid, vis) + 
               dfs(i, j - 1, grid, vis);
    }
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int n = grid.size(), m = grid[0].size();
        vector<vector<int>> vis(n, vector<int>(m, 0));
        int ans = 0;
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(grid[i][j] == 1 && vis[i][j] == 0) {
                    int cnt = dfs(i, j, grid, vis);
                    ans = max(ans, cnt);
                }
            }
        }
        return ans;
    }
};
